rfcs/0003-template-compiler.md

```diff
@@ -0,0 +1,126 @@
+---
+Status: Active
+Champions: @AndrewJakubowicz
+PR: https://github.com/lit/rfcs/pull/21
+---
+
+# @lit-labs/compiler
+
+A new labs package for preparing Lit templates at build time.
+
+## Objective
+
+Provide an optional build time performance improvement for the first render of a lit template.
+
+
+### Goals
+
+- Provide a build time transform that replaces lit `html` tagged template expressions with compiled templates.
```

Comment:
Recommend an additional goal:

It's a pure optimization. The code behaves exactly the same before and after, excepting code that introspects on TemplateResults.

Reply:
The only caveat to this goal is that the compiled template does not contain exactly the same marker comment contents (which is what I was referring to in the downside).

For example:

``` html`<div>${'hi'}<!--${''}--></div>` ``` when uncompiled results in the runtime prepared template of:

```
<div><!--?lit$983224318$--><!--lit$983224318$--></div>
```

But the compiled prepared html has the markers stripped:

```
<div><!--?--><!----></div>
```

I made this choice to reduce the size of the prepared compiled HTML since the markers are no longer needed once the template parts array is created. However we could keep the markers, or replace them with a placeholder marker such as `lit$54321$`.

Reply:
+1, I think these changes handle it well.